stackstate_checks_base/stackstate_checks/base/utils/http_helper.py

```diff
@@ -0,0 +1,1222 @@
+from schematics.exceptions import DataError
+from requests import Session, Request
+from requests.auth import HTTPBasicAuth
+from requests.adapters import HTTPAdapter
+from requests.packages.urllib3.util.retry import Retry
+
+try:
+    import urlparse
+except ImportError:
+    import urllib.parse as urlparse
+
+
+class HTTPBasicAuthentication(Model):
+    username = StringType(required=True)
+    password = StringType(required=True)
+
+
+class HTTPAuthenticationType(Enum):
+    BasicAuth = HTTPBasicAuthentication
+
+
+class HTTPMethod(Enum):
+    GET = 'GET'
+    POST = 'POST'
+    PUT = 'PUT'
+    PATCH = 'PATCH'
+    DELETE = 'DELETE'
+
+
+class HTTPRequestType(Enum):
+    JSON = dict
+
+
+class HTTPResponseType(Enum):
+    PLAIN = str
+    JSON = dict
+
+
+class HTTPHelperModel(Model):
+    """
+        This is the model for the `HTTPHelper` class.
+        When using the `HTTPHelper` class there will be the only acceptable values to create state and kick off the
+        request
+    """
+    mock_enable = BooleanType(required=False, default=False)
+    mock_status = IntType(required=False, default=None)
+    mock_response = BaseType(required=False, default=None)
+    endpoint = StringType(required=True, default=None)
+    auth_data = DictType(BaseType, required=False, default=None)
+    auth_type = BaseType(required=False, default=None)
+    session_auth_data = BaseType(required=False, default=None)
+    session_auth_type = BaseType(required=False, default=None)
+    body = BaseType(required=False, default=None)
+    proxy = BaseType(required=False, default=None)
+    headers = DictType(StringType, required=False, default=None)
+    session_headers = DictType(StringType, required=False, default=None)
+    query = DictType(StringType, required=False, default=None)
+    timeout = IntType(required=False, default=None)
+    ssl_verify = BooleanType(required=False, default=True)
+    retry_policy = BaseType(required=False, default=None)
+    request_schematic_validation = BaseType(required=False, default=None)
+    request_type_validation = BaseType(required=False, default=None)
+    response_status_code_validation = IntType(required=False, default=None)
+    response_type_validation = BaseType(required=False, default=None)
+    response_schematic_validation = BaseType(required=False, default=None)
+
+
+class HTTPHelper:
+    """
+        The HTTP Helper Handler is used to create compact methods using most of the function defined in the
+        Connection, Request, Session and Response Helpers
+        Functionality:
+            - GET Request
+            - POST Request
+            - DELETE Request
+            - PATCH Request
+            - PUT Request
+
+        All the HTTPHelper classes and Sub classes can accept a schematic dictionary that allows certain values to
+        manipulate state with
+    """
+
+    def __init__(self):
+        self.log = logging.getLogger('%s.base' % __name__)
+
+    def get(self, http_model):
```

Comment:
Perhaps it would be good to add another parameter here to pass a session handler for reuse purposes. Same for the other methods in this class.